pkg/provider/aws/secret.go

```diff
@@ -0,0 +1,319 @@
+// Copyright © 2024 Bank-Vaults Maintainers
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aws
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+	"github.com/aws/aws-sdk-go/service/ssm"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/bank-vaults/secrets-webhook/pkg/provider"
+	"github.com/bank-vaults/secrets-webhook/pkg/provider/common"
+)
+
+func (m *mutator) MutateSecret(ctx context.Context, mutateRequest provider.SecretMutateRequest) error {
+	// do an early exit if no mutation is needed
+	requiredToMutate, storeType, err := secretNeedsMutation(mutateRequest.Secret)
+	if err != nil {
+		return fmt.Errorf("checking if secret needs mutation failed: %w", err)
+	}
+
+	if !requiredToMutate {
+		return nil
+	}
+
+	err = m.newClient(ctx, mutateRequest.K8sClient)
+	if err != nil {
+		return fmt.Errorf("creating AWS clients failed: %w", err)
+	}
+
+	if value, ok := mutateRequest.Secret.Data[corev1.DockerConfigJsonKey]; ok {
+		var dc common.DockerCredentials
+		err := json.Unmarshal(value, &dc)
+		if err != nil {
+			return fmt.Errorf("unmarshal dockerconfig json failed: %w", err)
+		}
+
+		err = mutateDockerCreds(ctx, mutateRequest.Secret, *m.client, storeType, &dc)
+		if err != nil {
+			return fmt.Errorf("mutate docker creds failed: %w", err)
+		}
```

Comment:
Yes, because it is possible to have multiple secrets defined in a single resource.
We also have a test-case for this with Vault provider: <https://github.com/bank-vaults/secrets-webhook/blob/main/e2e/test/secret-vault.yaml>